src/server/pkg/storage/track/pg_tracker.go

```diff
@@ -0,0 +1,222 @@
+	var expiresAt time.Time
+	err := t.db.GetContext(ctx, &expiresAt,
+		`UPDATE storage.tracker_objects
+		SET expires_at = CURRENT_TIMESTAMP + $2 * interval '1 microsecond'
+		WHERE str_id LIKE $1 || '%'
+		RETURNING expires_at`, prefix, ttl.Microseconds())
+	if err != nil {
+		return time.Time{}, err
+	}
+	return expiresAt, nil
+}
+
+func (t *PostgresTracker) GetDownstream(ctx context.Context, id string) ([]string, error) {
+	dwn := []string{}
+	if err := t.db.SelectContext(ctx, &dwn,
+		`WITH target AS (
+			SELECT int_id FROM storage.tracker_objects WHERE str_id = $1
+		)
+		SELECT str_id
+		FROM storage.tracker_objects
+		WHERE int_id IN (
+			SELECT to_id FROM storage.tracker_refs WHERE from_id IN (SELECT int_id FROM target)
+		)`, id); err != nil {
+		return nil, err
+	}
+	return dwn, nil
+}
+
+func (t *PostgresTracker) GetUpstream(ctx context.Context, id string) ([]string, error) {
+	ups := []string{}
+	if err := t.db.SelectContext(ctx, &ups,
+		`WITH target AS (
+			SELECT int_id FROM storage.tracker_objects WHERE str_id = $1
+		)
+		SELECT str_id
+		FROM storage.tracker_objects
+		WHERE int_id IN (
+			SELECT from_id FROM storage.tracker_refs WHERE to_id IN (SELECT int_id FROM TARGET)
+		)`, id); err != nil {
+		return nil, err
+	}
+	return ups, nil
+}
+
+func (t *PostgresTracker) MarkTombstone(ctx context.Context, id string) error {
+	var tombstones []bool
+	if err := t.db.SelectContext(ctx, &tombstones, `	
+		UPDATE storage.tracker_objects
+		SET tombstone = (	
+			CASE
+				WHEN NOT EXISTS (
+					SELECT from_id FROM storage.tracker_refs
+					WHERE to_id IN (
+						SELECT int_id FROM storage.tracker_objects
+						WHERE str_id = $1
+					)
+				) THEN TRUE
+				ELSE FALSE
+			END
+		)
+		WHERE str_id = $1
+		RETURNING tombstone
+	`, id); err != nil {
+		return err
+	}
+	// if we get no results back, then it doesn't exist
+	if len(tombstones) == 0 {
+		return nil
+	}
+	// if the tombstone is not set, then it was because it would create a dangling ref
+	if !tombstones[0] {
+		return ErrDanglingRef
+	}
+	return nil
+}
+
+func (t *PostgresTracker) FinishDelete(ctx context.Context, id string) error {
+	err := t.withTx(ctx, func(tx *sqlx.Tx) error {
+		var tombstone bool
+		if err := tx.GetContext(ctx, &tombstone,
+			`DELETE FROM storage.tracker_objects
+			WHERE str_id = $1
+			RETURNING tombstone
+			`, id); err != nil {
+			return err
+		}
+		if !tombstone {
+			return ErrNotTombstone
+		}
+		if _, err := tx.ExecContext(ctx, `DELETE FROM storage.tracker_refs WHERE from_id IN (
+			SELECT int_id FROM storage.tracker_objects WHERE str_id = $1
+		)
+		`, id); err != nil {
+			return err
+		}
+		return nil
+	})
+	if err == sql.ErrNoRows {
+		return nil
+	}
+	return nil
+}
+
+func (t *PostgresTracker) IterateDeletable(ctx context.Context, cb func(id string) error) error {
+	rows, err := t.db.QueryxContext(ctx,
+		`SELECT str_id FROM storage.tracker_objects
+		WHERE (expires_at <= CURRENT_TIMESTAMP OR tombstone)
```

Comment:
We should still restructure this such that we first get the set of objects that are not referenced, then delete if the object is expired or there is a tombstone.